include/nbl/builtin/hlsl/bxdf/ndf.hlsl

```diff
@@ -16,53 +16,63 @@ namespace bxdf
 namespace ndf
 {
 
-namespace dummy_impl
+namespace impl
 {
-using sample_t = SLightSample<ray_dir_info::SBasic<float> >;
-using interaction_t = surface_interactions::SAnisotropic<surface_interactions::SIsotropic<ray_dir_info::SBasic<float> > >;
-using cache_t = SAnisotropicMicrofacetCache<SIsotropicMicrofacetCache<float> >;
-struct MetaQuery   // nonsense struct, just put in all the functions to pass the ndf query concepts
+template<typename T>
+struct NDFQuantQuery
 {
-    using scalar_type = float;
+    using scalar_type = T;
 
-    scalar_type getNdf() NBL_CONST_MEMBER_FUNC { return 0; }
-    scalar_type getLambdaL() NBL_CONST_MEMBER_FUNC { return 0; }
-    scalar_type getLambdaV() NBL_CONST_MEMBER_FUNC { return 0; }
+    scalar_type getVdotHLdotH() NBL_CONST_MEMBER_FUNC { return VdotHLdotH; }
+    scalar_type getVdotH_etaLdotH() NBL_CONST_MEMBER_FUNC { return VdotH_etaLdotH; }
 
-    scalar_type getG1over2NdotV() NBL_CONST_MEMBER_FUNC { return 0; }
-    scalar_type getOrientedEta() NBL_CONST_MEMBER_FUNC { return 0; }
-    scalar_type getDevshV() NBL_CONST_MEMBER_FUNC { return 0; }
-    scalar_type getDevshL() NBL_CONST_MEMBER_FUNC { return 0; }
-    BxDFClampMode getClampMode() NBL_CONST_MEMBER_FUNC { return BxDFClampMode::BCM_NONE; }
+    scalar_type VdotHLdotH;
+    scalar_type VdotH_etaLdotH;
 };
 }
 
+namespace dummy_impl
+{
+using sample_t = SLightSample<ray_dir_info::SBasic<float> >;
+using interaction_t = surface_interactions::SAnisotropic<surface_interactions::SIsotropic<ray_dir_info::SBasic<float> > >;
+using cache_t = SAnisotropicMicrofacetCache<SIsotropicMicrofacetCache<float> >;
+}
+
 #define NBL_CONCEPT_NAME NDF
 #define NBL_CONCEPT_TPLT_PRM_KINDS (typename)
 #define NBL_CONCEPT_TPLT_PRM_NAMES (T)
 #define NBL_CONCEPT_PARAM_0 (ndf, T)
-#define NBL_CONCEPT_PARAM_1 (query, dummy_impl::MetaQuery)
+#define NBL_CONCEPT_PARAM_1 (quant_query, typename T::quant_query_type)
 #define NBL_CONCEPT_PARAM_2 (_sample, dummy_impl::sample_t)
 #define NBL_CONCEPT_PARAM_3 (interaction, dummy_impl::interaction_t)
 #define NBL_CONCEPT_PARAM_4 (cache, dummy_impl::cache_t)
```

Comment:
shouldn't you take the `float` in the dummy_impl to be `T::scalar_type` or something from the NDF

Reply:
I'm not sure why but I'm unable to do
```
template<typename T>
using sample_t = SLightSample<ray_dir_info::SBasic<T> >;
```
and pass in `T::scalar_type` without the concept not working